pkg/sip/protocol.go

```diff
@@ -59,3 +63,132 @@ func sendAndACK(c Signaling, req *sip.Request) {
 		_ = c.WriteRequest(sip.NewAckRequest(req, r, nil))
 	}
 }
+
+func NewReferRequest(inviteRequest *sip.Request, inviteResponse *sip.Response, referToUrl string) *sip.Request {
+	req := sip.NewRequest(sip.REFER, inviteRequest.Recipient)
+
+	req.SipVersion = inviteRequest.SipVersion
+	sip.CopyHeaders("Via", inviteRequest, req)
+	// if inviteResponse.IsSuccess() {
+	// update branch, 2xx ACK is separate Tx
+	viaHop, _ := req.Via()
+	viaHop.Params.Add("branch", sip.GenerateBranch())
+	// }
+
+	if len(inviteRequest.GetHeaders("Route")) > 0 {
+		sip.CopyHeaders("Route", inviteRequest, req)
+	} else {
+		hdrs := inviteResponse.GetHeaders("Record-Route")
+		for i := len(hdrs) - 1; i >= 0; i-- {
+			rrh, ok := hdrs[i].(*sip.RecordRouteHeader)
+			if !ok {
+				continue
+			}
+
+			h := rrh.Clone()
+			req.AppendHeader(h)
+		}
+	}
+
+	maxForwardsHeader := sip.MaxForwardsHeader(70)
+	req.AppendHeader(&maxForwardsHeader)
+
+	if h, _ := inviteRequest.From(); h != nil {
+		sip.CopyHeaders("From", inviteRequest, req)
+	}
+
+	if h, _ := inviteResponse.To(); h != nil {
+		sip.CopyHeaders("To", inviteResponse, req)
+	}
+
+	if h, _ := inviteRequest.CallID(); h != nil {
+		sip.CopyHeaders("Call-ID", inviteRequest, req)
+	}
+
+	if h, _ := inviteRequest.CSeq(); h != nil {
+		sip.CopyHeaders("CSeq", inviteRequest, req)
+		// Increase the invite request cseq so that we use the proper cseq in the bye request eventually
+		h.SeqNo = h.SeqNo + 1
```

Comment:
This part is interesting. I assume BYE will also +1 it and will get the same CSeq as the REFER, right? Maybe we could set a correct CSeq on the BYE, without mutating the original invite?

Reply:
I changed it overwrite the CSeq of all requests we send out using a counter in the object. It's a bit more clear, but more code. Ideally, we'd fix sipgo to stop making assumptions.